test/distributed/fsdp/test_fsdp_mixed_precision.py

```diff
@@ -927,6 +927,97 @@ def test_float16_on_one_submodule_skip_inputs_error(self):
         ):
             fsdp(x).sum().backward()
 
+    @skip_if_lt_x_gpu(2)
+    def test_submodules_with_different_precisions_error(self):
+        forward_inputs: Dict[nn.Module, torch.Tensor] = {}
+        float16 = MixedPrecision(param_dtype=torch.float16, cast_forward_inputs=True)
+        float8 = MixedPrecision(param_dtype=torch.float8, cast_forward_inputs=True)
+
+        model = SaveForwardInputsModel(
+            forward_inputs=forward_inputs, cast_forward_inputs=False
+        ).cuda()
+        c1, c2 = model.c1, model.c2
+        x = torch.zeros(2, 100, device="cuda")
+
+        # For submodules with different precisions, right now current design
+        # does not support the case when the root FSDP instance wraps a submodule
+        # that is not the first one executed. Because for that submodule, its inputs
+        # (or previous submodule's outputs) have no way to be casted, instead,
+        # the root module's inputs are casted upfront before entering
+        # root module's forward
+        model.c1 = FSDP(model.c1, mixed_precision=float16)
+        fsdp = FSDP(model, mixed_precision=float8)
+        with self.assertRaisesRegex(
+            RuntimeError, "mat1 and mat2 must have the same dtype"
+        ):
+            fsdp(x).sum().backward()
+
+    @skip_if_lt_x_gpu(2)
+    def test_submodules_with_different_precisions(self):
+        forward_inputs: Dict[nn.Module, torch.Tensor] = {}
+        float16 = MixedPrecision(param_dtype=torch.float16, cast_forward_inputs=True)
+        float8 = MixedPrecision(param_dtype=torch.float8, cast_forward_inputs=True)
+
+        model = SaveForwardInputsModel(
+            forward_inputs=forward_inputs, cast_forward_inputs=False
+        ).cuda()
+        c1, c2 = model.c1, model.c2
+        x = torch.zeros(2, 100, device="cuda")
+
+        # For submodules with different precisions, it is recommended to wrap those
+        # submodules seperatedly with ``cast_forward_inputs``=True, for root FSDP instance,
```

Comment:
Should this be a new sentence?
```suggestion
        # submodules seperatedly with ``cast_forward_inputs``=True. For root FSDP instance,
```

 I am having some trouble following the explanation starting from "For root FSDP instance".

Can you explain what you mean by "make sure its wrapped submodules are executed before other submodules"? What is the reason for the order dependency? How can it be violated (given how modules are nested)?

> it is OK to set ``cast_forward_inputs`` or ``cast_root_forward_inputs`` to be True

This is just regarding the root FSDP instance right? Should we explain that we still prefer to set `cast_root_forward_inputs=True` in that case for logical clarity?

More broadly, should this comment be with the unit test, or should this be somewhere more visible?